### PR TITLE
8339132: Make DirectCodeBuilder write through without allocating instruction objects

### PR DESCRIPTION
**Pull request by liach**
Make `DirectCodeBuilder` write instructions actually directly without allocating extra objects. This speed up a lot of simple Class-File building cases that never go through intermediate transforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339132](https://bugs.openjdk.org/browse/JDK-8339132): Make DirectCodeBuilder write through without allocating instruction objects (**Sub-task** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) Review applies to [d798b302](https://git.openjdk.org/jdk/pull/20738/files/d798b302ded8b92fa83c639c7c8f0555ee064807)
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20738/head:pull/20738` \
`$ git checkout pull/20738`

Update a local copy of the PR: \
`$ git checkout pull/20738` \
`$ git pull https://git.openjdk.org/jdk.git pull/20738/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20738`

View PR using the GUI difftool: \
`$ git pr show -t 20738`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20738.diff">https://git.openjdk.org/jdk/pull/20738.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20738#issuecomment-2313934095)

### CODE REVIEW

**Review by cl4es — APPROVED** [review]
This seem to work great. A reduction on all relevant startup and footprint metrics. -26 classes on `StringConcatStartup`, for example.

Perhaps the code changes could be reduced somewhat if `DirectCodeBuilder::writeBytecode(Opcode)` et.c. returned `DirectCodeBuilder`.

File: src/java.base/share/classes/java/lang/classfile/CodeBuilder.java [hunk 1]
```diff
@@ -639,32 +639,34 @@ default CodeBuilder loadConstant(ConstantDesc value) {
...
+                    case 2 -> iconst_2();
+                    case 3 -> iconst_3();
+                    case 4 -> iconst_4();
+                    case 5 -> iconst_5();
+                    default -> (iVal >= Byte.MIN_VALUE && iVal <= Byte.MAX_VALUE) ? bipush(iVal)
+                            : (iVal >= Short.MIN_VALUE && iVal <= Short.MAX_VALUE) ? sipush(iVal)
+                            : ldc(constantPool().intEntry(iVal));
+                };
+            if (value instanceof Long lVal)
+                return lVal == 0l ? lconst_0()
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/classfile/CodeBuilder.java:657]:
```suggestion
                return lVal == 0L ? lconst_0()
```

File: src/java.base/share/classes/java/lang/classfile/CodeBuilder.java [hunk 2]
```diff
@@ -639,32 +639,34 @@ default CodeBuilder loadConstant(ConstantDesc value) {
...
+                        : lVal == 1l ? lconst_1()
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/classfile/CodeBuilder.java:658]:
```suggestion
                        : lVal == 1L ? lconst_1()
```

File: src/java.base/share/classes/java/lang/classfile/CodeBuilder.java [hunk 3]
```diff
@@ -639,32 +639,34 @@ default CodeBuilder loadConstant(ConstantDesc value) {
...
+                        : ldc(constantPool().longEntry(lVal));
+            if (value instanceof Float fVal)
+                return Float.floatToRawIntBits(fVal) == 0 ? fconst_0()
+                        : fVal == 1.0f ? fconst_1()
+                        : fVal == 2.0f ? fconst_2()
+                        : ldc(constantPool().floatEntry(fVal));
+            if (value instanceof Double dVal)
+                return Double.doubleToRawLongBits(dVal) == 0l ? dconst_0()
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/classfile/CodeBuilder.java:666]:
```suggestion
                return Double.doubleToRawLongBits(dVal) == 0L ? dconst_0()
```

**Review by cl4es — APPROVED** [review]

**Review by asotona — APPROVED** [review]
Looks good to me.

**Comment by liach** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**